heaps/min_heap.py

```diff
@@ -57,20 +67,51 @@ def heap_up(self, index):
             property is reestablished.
             
             This could be **very** helpful for the add method.
-            Time complexity: ?
-            Space complexity: ?
+            Time complexity: ? O(log n)
+            Space complexity: ? O(1)
         """
-        pass
+        parent_index = int((index - 1) / 2)
+        current_key = self.store[index].key 
+        parent_key = self.store[parent_index].key 
 
+        while current_key < parent_key:
+            self.swap(index, parent_index)
+            index = parent_index
+            parent_index = int((index - 1) / 2)
+            current_key = self.store[index].key
+            parent_key = self.store[parent_index].key 
+
+    def has_left_child(self, index):
+        left_child_index = int((2 * index) + 1)
+        if left_child_index <= len(self.store) - 1:
+            return True
+        return False
+
+    def has_right_child(self, index):
+        right_child_index = int((2 * index) + 2)
+        if right_child_index <= len(self.store) - 1:
+            return True
+        return False
+    
     def heap_down(self, index):
         """ This helper method takes an index and 
             moves the corresponding element down the heap if it's 
             larger than either of its children and continues until
             the heap property is reestablished.
+            time complexity O(log n)
+            space complexity O(1)
         """
-        pass
+        if self.has_left_child(index) and (self.store[index].key > self.store[int((2 * index) + 1)].key):
+            left_child_index = (2 * index) + 1
+            self.swap(index, left_child_index)
+            self.heap_down(left_child_index)
```

Comment:
Before deciding to swap here, we would also need to compare the left child to the right (if present). We only want to swap the one child which is the smallest.

Currently, if the left child were smaller then the parent, but larger than the right, first we would continue swapping the parent down the left subtree, then potentially need to swap the new parent down the right sub tree. Consider the arrangement [3, 2, 1]. 3 > 2 so it would swap with the left child → [2, 3, 1]. But notice the heap property is not re-established since 2 > 1. This code will continue and swap 2 and 1 → [1, 3, 2]. But if we had compare the values of the children first, we could see that 1 < 2 and swap that with 3 directly → [1, 2, 3] which re-establishes the heap in a single swap.